core/pkg/artifacts/saver.go

```diff
@@ -483,19 +484,24 @@ func (as *ArtifactSaver) uploadMultipart(
 	partEtags := make([]gql.UploadPartsInput, len(partData))
 
 	for t := range partResponses {
-		err := t.task.Err
+		uploadTask, ok := t.task.(*filetransfer.DefaultUploadTask)
+		if !ok {
+			err = fmt.Errorf("received unexpected task type %v", t.task.String())
+			return uploadResult{name: fileInfo.name, err: err}
+		}
```

Comment:
this check isn't ideal, but I don't think it can be avoided unless we 
1. use generics to make `CompletionCallback` act on the specific type (like `DefaultTask`), which leads to more issues in `file_transfer_manager` because this would also have to be generic to make `addTask` work
2. make `Response` a field on `Task`- this also doesn't make sense because we don't use this field for ref artifacts

Reply:
It would be better if instead of `SetCompletionCallback`/`GetCompletionCallback`, `Task` had a `Complete` method. Each implementing struct can accept a CompletionCallback of the correct type and implement `Complete` itself, e.g.:

```go
func (t *DefaultTask) SetCompletionCallback(cb func(*DefaultTask)) {
	t.completionCallback = cb
}

func (t *DefaultTask) Complete() {
	if t.completionCallback == nil {
		return
	}

	t.completionCallback(t)
}
```

But better yet, I don't think the callbacks need to accept a reference to the task, and you don't really need a setter method. So you could add this on all tasks using a mixin pattern:

```go
type taskCompletionCallback struct {
	CompletionCallback func()
}

func (t taskCompletionCallback) Complete() {
	if t.CompletionCallback == nil {
		return
	}

	t.CompletionCallback(t)
}

// By embedding taskCompletionCallback, DefaultTask
// and ReferenceArtifactTask gain the Complete method.

type DefaultTask struct {
	taskCompletionCallback
	...
}

type ReferenceArtifactTask struct {
	taskCompletionCallback
	...
}
```